### PR TITLE
feat: add support for discovering analysis options file from contex

### PR DESCRIPTION
**Pull request by incendial**
### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix
[ ] New rule
[ ] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[X] Add something to the core
[ ] Other, please explain:

### CODE REVIEW

File: lib/src/analyzers/lint_analyzer/lint_analyzer.dart [hunk 1]
```diff
@@ -78,24 +80,35 @@ class LintAnalyzer {
...
+      final analysisOptions = optionsFilePath == null
+          ? await analysisOptionsFromFilePath(rootFolder)
+          : await analysisOptionsFromFile(File(optionsFilePath));
+
+      final analysisOptionsConfig = ConfigBuilder.getConfig(analysisOptions);
+      final lintConfig =
+          ConfigBuilder.getLintConfig(analysisOptionsConfig, rootFolder)
+              .merge(config);
+
+      final filePaths = folders
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/lint_analyzer.dart:97]: @incendial please extract to separate function

**Review by dkrutskikh — APPROVED** [review]

File: analysis_options.yaml [hunk 2]
```diff
@@ -2,7 +2,8 @@ include: package:pedantic/analysis_options.yaml
 
 analyzer:
   exclude:
-    - test/resources/**
+    - test/resources/*
+    - test/resources/unused_files_analyzer/**
     - test/resources/unused_files_analyzer/generated/**/**
```
Comment (dkrutskikh) [inline analysis_options.yaml:7]: @incendial you can use only one `    - test/resources/unused_files_analyzer/**`

File: analysis_options.yaml [hunk 3]
```diff
@@ -20,7 +21,15 @@ dart_code_metrics:
     maximum-nesting: 5
     number-of-parameters: 4
   metrics-exclude:
-    - test/**
+    - test/analyzer_plugin/**
```
Comment (dkrutskikh) [inline analysis_options.yaml:24]: this is not scalable hell :(

File: lib/src/analyzers/lint_analyzer/lint_analysis_config.dart [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import 'package:glob/glob.dart';
+
+import 'anti_patterns/models/obsolete_pattern.dart';
+import 'metrics/models/metric.dart';
+import 'rules/models/rule.dart';
+
+class LintAnalysisConfig {
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/lint_analysis_config.dart:7]: @incendial please add `@immutable` annotation, add some documentation like:

This object represent merged configs and operates with our business objects.

File: lib/src/analyzers/lint_analyzer/lint_config.dart [hunk 5]
```diff
@@ -1,27 +1,63 @@
...
 
-import 'anti_patterns/models/obsolete_pattern.dart';
-import 'metrics/models/metric.dart';
-import 'rules/models/rule.dart';
+import '../../cli/models/parsed_arguments.dart';
+import '../../config_builder/analysis_options_utils.dart';
+import '../../config_builder/models/analysis_options.dart';
+import 'metrics/metrics_factory.dart';
 
+/// Class representing config
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/lint_config.dart:8]: @incendial please add information about this class is optimized for manipulation like merge

File: lib/src/analyzers/unused_files_analyzer/unused_files_analysis_config.dart [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import 'package:glob/glob.dart';
+
+class UnusedFilesAnalysisConfig {
```
Comment (dkrutskikh) [inline lib/src/analyzers/unused_files_analyzer/unused_files_analysis_config.dart:3]: @incendial please add documentation

File: lib/src/utils/exclude_utils.dart [hunk 7]
```diff
@@ -11,3 +11,12 @@ Iterable<Glob> prepareExcludes(Iterable<String> patterns, String root) =>
     patterns
         .map((exclude) => Glob(p.join(root, exclude).replaceAll(r'\', '/')))
         .toList();
+
+Iterable<Glob> prepareNormalizedExcludes(
```
Comment (dkrutskikh) [inline lib/src/utils/exclude_utils.dart:15]: @incendial after introduce prepareNormalizedExcludes, do we really need prepareExcludes ?

File: lib/src/utils/file_utils.dart [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import 'dart:io';
+
+import 'package:file/local.dart';
+import 'package:glob/glob.dart';
+import 'package:path/path.dart';
+
+import 'exclude_utils.dart';
+
+Set<String> extractFilesFromFolders(
```
Comment (dkrutskikh) [inline lib/src/utils/file_utils.dart:9]: @incendial please rename to extractDartFilesFromFolders

**Review by dkrutskikh — CHANGES_REQUESTED** [review]

File: analysis_options.yaml (hunk 2 above)
Comment (incendial) [reply analysis_options.yaml]: Removed.

File: lib/src/analyzers/lint_analyzer/lint_analysis_config.dart (hunk 4 above)
Comment (incendial) [reply lib/src/analyzers/lint_analyzer/lint_analysis_config.dart:11]: Done.

File: lib/src/analyzers/lint_analyzer/lint_config.dart (hunk 5 above)
Comment (incendial) [reply lib/src/analyzers/lint_analyzer/lint_config.dart]: Done.

File: lib/src/analyzers/unused_files_analyzer/unused_files_analysis_config.dart (hunk 6 above)
Comment (incendial) [reply lib/src/analyzers/unused_files_analyzer/unused_files_analysis_config.dart:4]: Done.

File: lib/src/utils/exclude_utils.dart (hunk 7 above)
Comment (incendial) [reply lib/src/utils/exclude_utils.dart]: Replaced `prepareExcludes` with `prepareNormalizedExcludes` and kept old name.

File: lib/src/utils/file_utils.dart (hunk 8 above)
Comment (incendial) [reply lib/src/utils/file_utils.dart]: Done.

**Review by dkrutskikh — APPROVED** [review]

**Pull request merged**